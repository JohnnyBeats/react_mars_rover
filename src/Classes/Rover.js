import DirectionEnum from "./DirectionEnum";
import CommandEnum from "./CommandEnum";

export class RoverEngine{
    constructor(coordinates, direction, commands, maxGrid, roverTeam){
        this.coordinates = coordinates;
        this.direction = direction;
        this.maxGrid = maxGrid;
        this.commands = commands;
        this.roverTeam = roverTeam || new RoverTeam();
        this.commandIndex = -1;
        this.commandsComplete = false;
        this.degreesRotation = 0;
        this.isActive = false;

        switch (direction) {
            case DirectionEnum.NORTH:
                this.degreesRotation = 0;
                break;
            case DirectionEnum.EAST:
                this.degreesRotation = 90;
                break;
            case DirectionEnum.SOUTH:
                this.degreesRotation = 180;
                break;
            case DirectionEnum.WEST:
                this.degreesRotation = 270;
                break;
            default:
                this.degreesRotation = 0;
                break;
        }

    }

    fellowRoverParking = (coordinates) => {
        return (this.roverTeam.getFellowRovers(this).filter(rover =>
            rover.coordinates[0] === coordinates[0] && rover.coordinates[1] === coordinates[1]).length > 0
        );
    };

    // Assumption - Grid numbers will never be negative and can't go over grid max size, if command is to cross grid it will ignore and run next command
    processCommand = (command) => {
        const moveCoordinates = [this.coordinates[0], this.coordinates[1]];

        switch (command) {
            case CommandEnum.MOVE:
                if(this.direction === DirectionEnum.NORTH && this.coordinates[1] < this.maxGrid[1]){
                    moveCoordinates[1] = moveCoordinates[1] + 1;
                }else if(this.direction === DirectionEnum.EAST && moveCoordinates[0] < this.maxGrid[0]){
                    moveCoordinates[0] = moveCoordinates[0] + 1;
                }else if(this.direction === DirectionEnum.SOUTH && moveCoordinates[1] > 0){
                    moveCoordinates[1] = moveCoordinates[1] - 1;
                }else if(this.direction === DirectionEnum.WEST && moveCoordinates[0] > 0){
                    moveCoordinates[0] = moveCoordinates[0] - 1;
                }
                break;
            case CommandEnum.LEFT:
                if(this.direction === DirectionEnum.NORTH){
                    this.direction = DirectionEnum.WEST;
                }else if(this.direction === DirectionEnum.EAST){
                    this.direction = DirectionEnum.NORTH;
                }else if(this.direction === DirectionEnum.SOUTH){
                    this.direction = DirectionEnum.EAST;
                }else if(this.direction === DirectionEnum.WEST){
                    this.direction = DirectionEnum.SOUTH;
                }
                this.degreesRotation = this.degreesRotation - 90;
                break;
            case CommandEnum.RIGHT:
                if(this.direction === DirectionEnum.NORTH){
                    this.direction = DirectionEnum.EAST;
                }else if(this.direction === DirectionEnum.EAST){
                    this.direction = DirectionEnum.SOUTH;
                }else if(this.direction === DirectionEnum.SOUTH){
                    this.direction = DirectionEnum.WEST;
                }else if(this.direction === DirectionEnum.WEST){
                    this.direction = DirectionEnum.NORTH;
                }

                this.degreesRotation = this.degreesRotation + 90;

                break;
            default:
                if(this.commandIndex < this.commands.length){
                    this.runNextCommand();
                }else{
                    this.commandsComplete = true;
                    this.isActive = false;
                }
        }

        if(moveCoordinates[0] !== this.coordinates[0] || moveCoordinates[1] !== this.coordinates[1]){
            //Can't be on the same block with another rover
            if(!this.fellowRoverParking(moveCoordinates)){
                this.coordinates = moveCoordinates;
            }
        }

    };

    runNextCommand = () => {
        if(!this.commandsComplete){
            this.isActive = true;
            this.commandIndex++;
            const command = this.commands[this.commandIndex];
            this.processCommand(command);
        }
    };

    getDirection = () => {
        return this.direction;
    };

    getPosition = () => {
        return this.coordinates[0] + " " +  this.coordinates[1] + " " + this.direction;
    };

    checkPosition = (positionCoordinates) => {
        return this.coordinates[0] === positionCoordinates[0] && this.coordinates[1] === positionCoordinates[1];
    };
}

export class RoverEngineMultiplayer extends RoverEngine{
    constructor(coordinates, direction, commands, maxGrid, roverTeam, id, name){
        super(coordinates, direction, commands, maxGrid, roverTeam);
        this.id = id;
        this.name = name;
    }
}

export class RoverTeam{
    constructor(){
        this.rovers = [];
    }

    getFellowRovers = (rover) => {
       return this.rovers.filter(fellow => rover !== fellow)
    }
}

export default {RoverEngine, RoverEngineMultiplayer, RoverTeam};