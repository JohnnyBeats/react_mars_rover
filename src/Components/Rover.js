import React from "react";
import DirectionEnum from "../Classes/DirectionEnum";

class Rover extends React.Component {
    render() {
        function getDirectionClassName(direction){
            const roverClassName = "rover";

            switch (direction) {
                case DirectionEnum.NORTH:
                    return roverClassName + " north";
                case DirectionEnum.EAST:
                    return roverClassName + " east";
                case DirectionEnum.SOUTH:
                    return roverClassName + " south";
                case DirectionEnum.WEST:
                    return roverClassName + " west";
                default:
                    return roverClassName;
            }
        }

        return (
            //Added inline style item here to ensure turning is smooth between directions
            <div data-testid="rover" className={getDirectionClassName(this.props.engine.getDirection()) + (this.props.engine.isActive ? " active" : " inActive")} style={{transform: 'rotate(' + this.props.engine.degreesRotation + 'deg)'}}>

            </div>
        )
    }
}

export default Rover;