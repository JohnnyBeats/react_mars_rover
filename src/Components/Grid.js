import React from "react";
import Rover from './Rover.js'

class Column extends React.Component {
    render() {
        const roverEngine = this.props.roverTeam.rovers.filter((item) => {return item.checkPosition([this.props.column, this.props.row])})[0];
        const coordinates = this.props.column + "," +this.props.row;

        return (
            <div className="column">
                {roverEngine !== undefined ?
                    <Rover engine={roverEngine}/> :
                    <span className="columnCoordinate" title={coordinates}>{coordinates}</span>}
            </div>
        )
    }
}

class Row extends React.Component {
    render() {
        return (
            <div className="row">
                {Array.from(Array(this.props.columns).keys()).map((value, index) => {
                    return <Column key={"column-" + index}  row={this.props.row} column={index} roverTeam={this.props.roverTeam}/>
                })}
            </div>
        )
    }
}

class Grid extends React.Component {
    render() {
        const maxX = this.props.gridSize[0];
        const maxY = this.props.gridSize[1];
        //const pageSize = 5;

        //we want to add one extra because we work on a zero based number
        let rows = maxY + 1;
        let columns = maxX + 1;

        //we need to page the grid because we will break the DOM with too many items
        /*if(rows > pageSize){
            rows = pageSize;
            columns = pageSize;
        }*/

        return (
            <div className={this.props.showCover ? "grid hidden": "grid"}>
                {Array.from(Array(rows).keys()).map((value, index) => {
                    return <Row  key={"row-" + index} columns={columns} row={maxY - index} roverTeam={this.props.roverTeam}/>
                })}
            </div>
        )
    }
}

export default Grid;