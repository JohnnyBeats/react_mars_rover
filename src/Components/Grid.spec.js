import React from 'react';
import { render } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import Grid from './Grid';
import {RoverEngine, RoverTeam} from "../Classes/Rover.js";

describe('The grid component', ()=> {
    const testData = () => {
        const maxGrid = [5,5];

        return {
            maxGrid
        }
    };

    it('renders 5 rows and 5 columns without regression', ()=>{
        const {maxGrid} = testData();
        const roverTeam = new RoverTeam();
        const {container} = render(
            <Grid showCover="false" roverTeam={roverTeam} gridSize={maxGrid}/>
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    it('displays a rover on 1 2 N', ()=>{
        const {maxGrid} = testData();
        const roverTeam = new RoverTeam();
        roverTeam.rovers.push(new RoverEngine([1, 2], "N", "M", maxGrid, roverTeam));

        const {container} = render(
            <Grid showCover="false" roverTeam={roverTeam} gridSize={maxGrid}/>
        );

        expect(container.firstChild).toMatchSnapshot();
    });
});