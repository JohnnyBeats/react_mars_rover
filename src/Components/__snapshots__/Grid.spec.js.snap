// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The grid component displays a rover on 1 2 N 1`] = `
<div
  class="grid hidden"
>
  <div
    class="row"
  >
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="0,5"
      >
        0,5
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="1,5"
      >
        1,5
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="2,5"
      >
        2,5
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="3,5"
      >
        3,5
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="4,5"
      >
        4,5
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="5,5"
      >
        5,5
      </span>
    </div>
  </div>
  <div
    class="row"
  >
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="0,4"
      >
        0,4
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="1,4"
      >
        1,4
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="2,4"
      >
        2,4
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="3,4"
      >
        3,4
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="4,4"
      >
        4,4
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="5,4"
      >
        5,4
      </span>
    </div>
  </div>
  <div
    class="row"
  >
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="0,3"
      >
        0,3
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="1,3"
      >
        1,3
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="2,3"
      >
        2,3
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="3,3"
      >
        3,3
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="4,3"
      >
        4,3
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="5,3"
      >
        5,3
      </span>
    </div>
  </div>
  <div
    class="row"
  >
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="0,2"
      >
        0,2
      </span>
    </div>
    <div
      class="column"
    >
      <div
        class="rover north inActive"
        data-testid="rover"
        style="transform: rotate(0deg);"
      />
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="2,2"
      >
        2,2
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="3,2"
      >
        3,2
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="4,2"
      >
        4,2
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="5,2"
      >
        5,2
      </span>
    </div>
  </div>
  <div
    class="row"
  >
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="0,1"
      >
        0,1
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="1,1"
      >
        1,1
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="2,1"
      >
        2,1
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="3,1"
      >
        3,1
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="4,1"
      >
        4,1
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="5,1"
      >
        5,1
      </span>
    </div>
  </div>
  <div
    class="row"
  >
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="0,0"
      >
        0,0
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="1,0"
      >
        1,0
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="2,0"
      >
        2,0
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="3,0"
      >
        3,0
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="4,0"
      >
        4,0
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="5,0"
      >
        5,0
      </span>
    </div>
  </div>
</div>
`;

exports[`The grid component renders 5 rows and 5 columns without regression 1`] = `
<div
  class="grid hidden"
>
  <div
    class="row"
  >
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="0,5"
      >
        0,5
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="1,5"
      >
        1,5
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="2,5"
      >
        2,5
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="3,5"
      >
        3,5
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="4,5"
      >
        4,5
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="5,5"
      >
        5,5
      </span>
    </div>
  </div>
  <div
    class="row"
  >
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="0,4"
      >
        0,4
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="1,4"
      >
        1,4
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="2,4"
      >
        2,4
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="3,4"
      >
        3,4
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="4,4"
      >
        4,4
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="5,4"
      >
        5,4
      </span>
    </div>
  </div>
  <div
    class="row"
  >
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="0,3"
      >
        0,3
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="1,3"
      >
        1,3
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="2,3"
      >
        2,3
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="3,3"
      >
        3,3
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="4,3"
      >
        4,3
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="5,3"
      >
        5,3
      </span>
    </div>
  </div>
  <div
    class="row"
  >
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="0,2"
      >
        0,2
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="1,2"
      >
        1,2
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="2,2"
      >
        2,2
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="3,2"
      >
        3,2
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="4,2"
      >
        4,2
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="5,2"
      >
        5,2
      </span>
    </div>
  </div>
  <div
    class="row"
  >
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="0,1"
      >
        0,1
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="1,1"
      >
        1,1
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="2,1"
      >
        2,1
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="3,1"
      >
        3,1
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="4,1"
      >
        4,1
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="5,1"
      >
        5,1
      </span>
    </div>
  </div>
  <div
    class="row"
  >
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="0,0"
      >
        0,0
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="1,0"
      >
        1,0
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="2,0"
      >
        2,0
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="3,0"
      >
        3,0
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="4,0"
      >
        4,0
      </span>
    </div>
    <div
      class="column"
    >
      <span
        class="columnCoordinate"
        title="5,0"
      >
        5,0
      </span>
    </div>
  </div>
</div>
`;
